src/packet_router/mod.rs

```diff
@@ -192,22 +237,45 @@ impl PacketRouter {
         Ok(())
     }
 
-    async fn send_packet(&mut self, packet: &CacheMessage<PacketUp>) -> Result {
+    async fn send_packet(
+        &mut self,
+        packet: &CacheMessage<PacketUp>,
+        keypair: Arc<Keypair>,
+    ) -> Result {
         debug!(packet_hash = packet.hash().to_b64(), "sending packet");
 
-        let uplink = mk_uplink(packet, self.keypair.clone()).await?;
+        let uplink = mk_uplink(packet, keypair).await?;
         self.service.send(uplink).await
     }
 }
 
 pub async fn mk_uplink(
     packet: &CacheMessage<PacketUp>,
     keypair: Arc<Keypair>,
-) -> Result<PacketRouterPacketUpV1> {
+) -> Result<envelope_up_v1::Data> {
     use std::ops::Deref;
     let mut uplink: PacketRouterPacketUpV1 = packet.deref().into();
     uplink.hold_time = packet.hold_time().as_millis() as u64;
     uplink.gateway = keypair.public_key().into();
     uplink.signature = sign(keypair, uplink.encode_to_vec()).await?;
-    Ok(uplink)
+    let envelope = envelope_up_v1::Data::Packet(uplink);
+    Ok(envelope)
+}
+
+pub async fn mk_session_key_init(
+    keypair: Arc<Keypair>,
+    offer: &PacketRouterSessionOfferV1,
+) -> Result<(Arc<Keypair>, envelope_up_v1::Data)> {
+    let session_keypair = Arc::new(mk_session_keypair());
+    let session_key = session_keypair.public_key();
+
+    let mut session_init = PacketRouterSessionInitV1 {
+        gateway: keypair.public_key().into(),
+        session_key: session_key.into(),
+        nonce: offer.nonce.clone(),
+        signature: vec![],
+    };
+    session_init.signature = sign(session_keypair.clone(), session_init.encode_to_vec()).await?;
```

Comment:
I'm not sure this is being signed by the right key. It should be signed by the gateway's registered keypair (e.g. the ECC608), not by the session key itself.

Reply:
Good catch... Fixed in [4767a67](https://github.com/helium/gateway-rs/pull/438/commits/4767a67d3148140d1482a722487f02f2ae264e05)